File: Obj-c/pamservices/TrackingData.m
//
//  TrackingData.m
//  pamservices
//
//  Created by heart on 9/8/2560 BE.
//  Copyright © 2560 heart. All rights reserved.
//

#import "TrackingData.h"
#import "FingerPrint.h"
#import "PAM.h"

@implementation TrackingData

-(TrackingData*)init{
    if( self = [super init] ){
        self.platform = @"iOS";
        self.do_not_track = @"unspecified";
        self.adblock = @"false";
        return self;
    }
    return nil;
}

-(NSString*)getFingerPrint{
    
    NSString *str = [NSString stringWithFormat:@"%@%@%@%@%@%@%@",
                    self.updfh,
                    self.appId,
                    self.page_language,
                    self.resolution,
                    self.timezone_offset ,
                    self.platform,
                    self.adblock
                    ];
   
    if(self.page_title != nil){
        str = [str stringByAppendingString:self.page_title];
    }
    
    if(self.page_referrer != nil){
        str = [str stringByAppendingString:self.page_title];
    }
    
    if(self.page_url != nil){
        str = [str stringByAppendingString:self.page_title];
    }
    
    if(self.counter != nil){
        str = [str stringByAppendingString:self.page_title];
    }
    
    return [FingerPrint fromString:str];
}

-(NSDictionary*)toDictionary{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    
    [dict setObject:self.appId forKey:@"app_id"];
    [dict setObject:self.page_language forKey:@"page_language"];
    [dict setObject:self.resolution forKey:@"resolution"];
    [dict setObject:self.timezone_offset forKey:@"timezone_offset"];
    [dict setObject:self.platform forKey:@"platform"];
    [dict setObject:self.do_not_track forKey:@"do_not_track"];
    [dict setObject:self.adblock forKey:@"adblock"];
    
    if(self.updfh != nil) {
       [dict setObject:self.updfh forKey:@"updfh"];
    }
    
    if(self.page_title != nil){
        [dict setObject:self.page_title forKey:@"page_title"];
    }
    
    if(self.page_referrer != nil){
        [dict setObject:self.page_referrer forKey:@"page_referrer"];
    }
    
    if(self.page_url != nil){
        [dict setObject:self.page_url forKey:@"page_url"];
        [PAM defaultTrackingData].page_referrer = self.page_url;
    }
    
    if(self.counter != nil){
        [dict setObject:self.counter forKey:@"counter"];
    }
    
    if(self.mtc_id != nil){
        [dict setObject:self.mtc_id forKey:@"mtc_id"];
    }
    
    if(self.sid != nil){
        [dict setObject:self.sid forKey:@"sid"];
    }
    
    [dict setObject:[self getFingerPrint] forKey:@"fingerprint"];
    
    return [NSDictionary dictionaryWithDictionary:dict];
}

@end

File: Swift/pamservices/TrackingData.swift
//
//  TrackingData.swift
//  pamservices
//
//  Created by narongrit kanhanoi on 9/8/2560 BE.
//  Copyright © 2017 pushandmotion.com. All rights reserved.
//


public class TrackingData {
    
    var page_title:String!
    var page_language:String!
    var page_referrer:String!
    var page_url:String!
    var resolution:String!
    var updfh:String!
    var appId:String!
    var mtc_id:String!
    var sid:String!
    var counter:Int!
    var timezone_offset:Int!
    var platform:String!
    var adblock:String!
    var do_not_track:String!
    
    init() {
        adblock = "false"
        do_not_track = "false"
    }
    
    public func getFingerPrint() -> String {
        var str: String = "\(updfh)\(appId)\(page_language)\(resolution)\(timezone_offset)\(platform)\(adblock)"
        if page_title != nil {
            str = str + (page_title)
        }
        if page_referrer != nil {
            str = str + (page_title)
        }
        if page_url != nil {
            str = str + (page_title)
        }
        if counter != nil {
            str = str + (page_title)
        }
        return FingerPrint.from(str)!
    }
    
    public func toDictionary() -> [AnyHashable : Any]! {
        var dict = [AnyHashable: Any]()
        dict["app_id"] = appId
        dict["page_language"] = page_language
        dict["resolution"] = resolution
        dict["timezone_offset"] = timezone_offset
        dict["platform"] = platform
        dict["do_not_track"] = do_not_track
        dict["adblock"] = "false"
        if updfh != nil {
            dict["updfh"] = updfh
        }
        if page_title != nil {
            dict["page_title"] = page_title
        }
        if page_referrer != nil {
            dict["page_referrer"] = page_referrer
        }
        if page_url != nil {
            dict["page_url"] = page_url
            PAM.defaultTrackingData().page_referrer = page_url
        }
        if counter != nil {
            dict["counter"] = counter
        }
        if mtc_id != nil {
            dict["mtc_id"] = mtc_id
        }
        if sid != nil {
            dict["sid"] = sid
        }
        dict["fingerprint"] = getFingerPrint()
        return dict
    }
}
